File: reference/typescript/src/transport/transport.ts
import type { AuditEvent } from '../schema/event.js';
import type { AuditCapability, NegotiationResult } from '../schema/capability.js';

// The exchange of §6, independent of the MCP wire: an attempt answered by an Attempt Response, and
// an outcome answered by nothing. A binding (§6.4, §6.5) carries both; nothing above this interface
// depends on which.

// Tier-1 reject reason codes (§7.6); the wire reason is pinned to this closed set.
export type RejectReason = 'schema-invalid' | 'replay-detected' | 'signature-invalid' | 'l2-unsigned' | 'unknown-key';

// accept = the record is sealed; proceed. reject = the record will not be sealed; do not proceed.
// unavailable = nothing was decided; do not proceed, and the identical attempt may be sent again
// (§7.1). None authorizes the domain action: fail-closed governs the record, not the action.
export type AttemptResponse =
  | {
      status: 'accept';
      seq: number;
      record_hash: string;
      host_ts: string;
      previous_hash: string;
      // The countersignature triple appears together or not at all (§7.1). A host declaring `none`
      // returns none of it; one declaring `host` returns all of it on every accept (§5.2).
      host_signature?: string;
      host_key_id?: string;
      log_id?: string;
    }
  | { status: 'reject'; reason: RejectReason }
  | { status: 'unavailable'; reason: 'internal-error' };

export interface AuditTransport {
  // The §6.1 comparison, for the call this transport serves.
  negotiate(offered: AuditCapability): NegotiationResult;

  // Resolves with the host's answer; bounding the wait is the binding's concern (§6).
  sendAttempt(event: AuditEvent): Promise<AttemptResponse>;

  sendOutcome(event: AuditEvent): Promise<void>;
}

File: reference/python/src/auditable_mcp/transport.py
"""Audit transport protocol interface.

Defines the request/response pattern for attempts, outcomes, and capability negotiation.
"""

from dataclasses import asdict, dataclass
from typing import Protocol

from auditable_mcp.capability import AuditCapability, NegotiationResult


@dataclass(frozen=True)
class AttemptResponse:
    """The host's answer to an attempt (§7.1).

    - `accept`: the record is sealed.
    - `reject`: the record will not be sealed.
    - `unavailable`: nothing was decided; the identical attempt may be sent again (§7.1).
    """

    status: str  # 'accept' | 'reject' | 'unavailable'
    seq: int | None = None
    record_hash: str | None = None
    # host_ts and previous_hash let the tool recompute the record hash and self-verify the seal.
    host_ts: str | None = None
    previous_hash: str | None = None
    reason: str | None = None
    # The countersignature triple appears together or not at all (§7.1). A host declaring `none`
    # returns none of it; one declaring `host` returns all of it on every accept (§5.2).
    host_signature: str | None = None
    host_key_id: str | None = None
    log_id: str | None = None


class AuditTransportError(Exception):
    """The transport could not carry an attempt or its answer: a fault or a protocol error (§6)."""


def response_members(response: AttemptResponse) -> dict:
    """Return the members an Attempt Response carries on the wire: the fields that are set."""
    return {name: value for name, value in asdict(response).items() if value is not None}


def accept(
    seq: int,
    record_hash: str,
    host_ts: str,
    previous_hash: str,
    host_signature: str | None = None,
    host_key_id: str | None = None,
    log_id: str | None = None,
) -> AttemptResponse:
    """Build an accept carrying what the tool needs to recompute the hash, and the countersignature if any."""
    return AttemptResponse(
        status='accept',
        seq=seq,
        record_hash=record_hash,
        host_ts=host_ts,
        previous_hash=previous_hash,
        host_signature=host_signature,
        host_key_id=host_key_id,
        log_id=log_id,
    )


def reject(reason: str) -> AttemptResponse:
    """Build a reject response."""
    return AttemptResponse(status='reject', reason=reason)


def unavailable(reason: str) -> AttemptResponse:
    """Build an unavailable response."""
    return AttemptResponse(status='unavailable', reason=reason)


class AuditTransport(Protocol):
    """Transport between a tool and the host audit subsystem."""

    def negotiate(self, offered: AuditCapability) -> NegotiationResult:
        """Exchange capabilities: present the tool's offer, receive the host requirement and fit."""
        ...

    def send_attempt(self, event: dict) -> AttemptResponse:
        """Send an attempt and block for the host's answer (§6).

        Raises:
            AuditTransportError: No answer came: a transport fault or a protocol error.
            OSError: The transport's I/O failed.
            TimeoutError: The bound on the wait elapsed.
        """
        ...

    def send_outcome(self, event: dict) -> None:
        """Send an outcome, which has no answer (§6)."""
        ...
